src/lib.rs

```diff
@@ -1,62 +1,143 @@
 extern crate openvr_sys;
 
-use openvr_sys::EVRInitError::*;
-use openvr_sys::EVRApplicationType::*;
-
-pub mod common;
-pub mod error;
-pub mod tracking;
-pub mod system;
-pub mod extended_display;
-pub mod compositor;
-pub mod render_models;
-pub mod tracked_camera;
-pub mod subsystems;
-
-pub use system::IVRSystem;
-pub use extended_display::IVRExtendedDisplay;
-pub use compositor::IVRCompositor;
-pub use render_models::IVRRenderModels;
-pub use tracked_camera::IVRTrackedCamera;
-
-pub use subsystems::*;
-pub use error::*;
-
-pub use common::Eye;
-
-/// Inits the open vr interface and returns the system
-pub fn init() ->  Result<system::IVRSystem, Error<openvr_sys::EVRInitError>> {
-    let mut err = EVRInitError_VRInitError_None;
-    let app_type = EVRApplicationType_VRApplication_Scene;
-
-    // try to initialize base vr eco
-    unsafe {
-        openvr_sys::VR_InitInternal(&mut err, app_type);
-    };
-
-    // check for errors
-    match err {
-        EVRInitError_VRInitError_None => {
-            // get system
-            let result = system();
-            match result {
-                Ok(sys) => {
-                    return Ok(sys);
-                },
-                Err(err) => {
-                    return Err(err);
-                }
+use std::sync::atomic::{Ordering, AtomicBool, ATOMIC_BOOL_INIT};
+use std::{fmt, error, slice};
+use std::ffi::CStr;
+
+use openvr_sys as sys;
+
+mod tracking;
+
+mod system;
+mod compositor;
+
+pub use tracking::*;
+
+static INITIALIZED: AtomicBool = ATOMIC_BOOL_INIT;
+
+/// Initialize OpenVR
+///
+/// # Panics
+/// When the library has already been initialized
+pub fn init(ty: ApplicationType) -> Result<Context, InitError> {
+    if INITIALIZED.swap(true, Ordering::AcqRel) {
+        panic!("OpenVR has already been initialized!");
+    }
+
+    let mut error = sys::EVRInitError_EVRInitError_VRInitError_None;
+    unsafe { sys::VR_InitInternal(&mut error, ty as sys::EVRApplicationType) };
+    if error != sys::EVRInitError_EVRInitError_VRInitError_None {
+        return Err(InitError(error));
+    }
+    if !unsafe { sys::VR_IsInterfaceVersionValid(sys::IVRSystem_Version.as_ptr() as *const i8) } {
+        unsafe { sys::VR_ShutdownInternal() }
+        return Err(InitError(sys::EVRInitError_EVRInitError_VRInitError_Init_InterfaceNotFound));
+    }
+    Ok(unsafe { Context::new() }?)
+}
+
+pub struct System<'a>(&'a sys::VR_IVRSystem_FnTable);
+pub struct Compositor<'a>(&'a sys::VR_IVRCompositor_FnTable);
+
+/// Entry points into OpenVR.
+///
+/// At most one of this object may exist at a time.
+pub struct Context {
```

Comment:
What I don't like on this approach is that you are forcing the user to instantiate all interfaces / subsystems.
If one of them fails the whole Context will be invalid. I think depending on the ApplicationType some interfaces are not available in will cause an error. The interfaces should be independent from each other.
As a quick solution we'd at least wrap an Option around each interface and make them all optional.